File: 19.powx-n/PowXN.cpp
#include <iostream>
#include <float.h>

using namespace std;

class Solution {
	public:
		double myPow(double x, int n) {
			if(n == 0 || x == 1.0) {
				return 1;
			}
			if(x == -1.0) {
				return n%2==1?-1.0:1.0;
			}
			long num = n;
			if(num < 0) {
				x = 1 / x;
				num = -num;
			}
			long double result = x;
			for(int i = 1; i < num; i ++) {
				result *= x;
				if(result >0 && result < DBL_MIN) {
					return 0;
				}
			}
			return result;
		}
};

int main(int argc, char *argv[])
{
	Solution s;
	cout << s.myPow(-1.0, 300000000);
	//cout << DBL_MIN;
	return 0;
}

File: 19.powx-n/PowXN.py
# 递归降幂

class Solution(object):
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        if n == 0 or x == 1.0:
            return 1.0
        if x == -1.0:
            return -1.0 if n%2==1 else 1.0
        if n < 0:
            return 1 / self.myPow(x, -n)
        if n % 2 == 0:
            return self.myPow(x * x, n / 2)
        else:
            return x * self.myPow(x * x, (n - 1) / 2)

if __name__ == "__main__":
    s = Solution();
    print(s.myPow(0.00001, 21472737712))
